### PR TITLE
NCL-8238 Integrate WebSocket - Builds

### PR DESCRIPTION
**Pull request by matedo1**

### CODE REVIEW

File: src/components/ArtifactUsagesPage/ArtifactUsagesPage.tsx [hunk 1]
```diff
@@ -1,25 +1,53 @@
...
+    useCallback(
+      (wsData: any) => {
+        if (hasBuildStarted(wsData)) {
+          serviceContainerBuildsRunner({
+            serviceData: { id: artifactId },
+            requestConfig: { params: componentQueryParamsObject },
+          });
+        } else if (hasBuildStatusChanged(wsData)) {
+          const wsBuild: Build = wsData.build;
+          serviceContainerBuildsSetter((previousBuildPage) => refreshPage<BuildPage, Build>(previousBuildPage!, wsBuild));
```
Comment (patrikk0123) [inline src/components/ArtifactUsagesPage/ArtifactUsagesPage.tsx:45]: `refreshPage(...)` is enough and types are infered

File: src/components/ArtifactUsagesPage/ArtifactUsagesPage.tsx [hunk 2]
```diff
@@ -1,25 +1,53 @@
...
+import { refreshPage } from 'utils/refreshHelper';
+
 interface IArtifactUsagesPageProps {
   componentId?: string;
 }
 
 export const ArtifactUsagesPage = ({ componentId = 'b1' }: IArtifactUsagesPageProps) => {
   const { artifactId } = useParamsRequired();
 
+  const { componentQueryParamsObject } = useComponentQueryParams(componentId);
```
Comment (patrikk0123) [inline src/components/ArtifactUsagesPage/ArtifactUsagesPage.tsx:24]: `artifactUsagesQueryParamsObject` (to keep is consistent)

File: src/components/BuildsPage/BuildsPage.tsx [hunk 3]
```diff
@@ -9,14 +15,37 @@ import { PageLayout } from 'components/PageLayout/PageLayout';
...
+
 interface IBuildsPageProps {
   componentId?: string;
 }
 
 export const BuildsPage = ({ componentId = 'b1' }: IBuildsPageProps) => {
   const serviceContainerBuilds = useServiceContainer(buildApi.getBuilds);
 
-  useQueryParamsEffect(serviceContainerBuilds.run, { componentId });
+  const { componentQueryParamsObject } = useComponentQueryParams(componentId);
```
Comment (patrikk0123) [inline src/components/BuildsPage/BuildsPage.tsx:27]: `buildsQueryParamsObject`

File: src/components/GroupBuildsPage/GroupBuildsPage.tsx [hunk 4]
```diff
@@ -9,10 +15,33 @@ import { PageLayout } from 'components/PageLayout/PageLayout';
 
 import * as groupBuildApi from 'services/groupBuildApi';
 
+import { refreshPage } from 'utils/refreshHelper';
+
 export const GroupBuildsPage = ({ componentId = 'gb1' }) => {
   const serviceContainerGroupBuilds = useServiceContainer(groupBuildApi.getGroupBuilds);
 
-  useQueryParamsEffect(serviceContainerGroupBuilds.run, { componentId });
+  const { componentQueryParamsObject } = useComponentQueryParams(componentId);
```
Comment (patrikk0123) [inline src/components/GroupBuildsPage/GroupBuildsPage.tsx:23]: `groupBuildsQueryParamsObject`

File: src/components/ProductMilestoneBuildsPerformedPage/ProductMilestoneBuildsPerformedPage.tsx [hunk 5]
```diff
@@ -1,25 +1,53 @@
...
+import { refreshPage } from 'utils/refreshHelper';
+
 interface IProductMilestoneBuildsPerformedPageProps {
   componentId?: string;
 }
 
 export const ProductMilestoneBuildsPerformedPage = ({ componentId = 'b1' }: IProductMilestoneBuildsPerformedPageProps) => {
   const { productMilestoneId } = useParamsRequired();
 
+  const { componentQueryParamsObject } = useComponentQueryParams(componentId);
```
Comment (patrikk0123) [inline src/components/ProductMilestoneBuildsPerformedPage/ProductMilestoneBuildsPerformedPage.tsx:24]: `buildPerformedQueryParamsObject`

**Review by patrikk0123 — COMMENTED** [review]
Also please create separate commits for significant changes in hooks and utils dirs.

File: src/components/ProductMilestoneBuildsPerformedPage/ProductMilestoneBuildsPerformedPage.tsx (hunk 5 above)
Comment (matedo1) [reply src/components/ProductMilestoneBuildsPerformedPage/ProductMilestoneBuildsPerformedPage.tsx:24]: I used `buildsPerformedQueryParamsObject`

File: src/utils/refreshHelper.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * @param originalPage - Original Page
+ * @param newItem - Item inside original Page will be replaced with new Item with the same ID
+ * @returns Refreshed Page containing new Item instead of the original Item with the same ID, when newItem is invalid, then Original Page is returned
+ */
+export const refreshPage = <T extends TPage, V extends TItem>(originalPage: T, newItem: V): T => {
+  if (newItem === undefined) {
+    uiLogger.error('refreshPage: invalid newItem (undefined value)');
+    return originalPage;
+  }
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/utils/refreshHelper.ts:20]: Once wsData.<entity> is passed from backend, there no constrain preventing backend from sending wrong data.

**Comment by matedo1** [comment]
See also: https://github.com/project-ncl/pnc-web-ui-react/pull/380/commits/c58b037db0a7ebcd86c75f2aa5aafa8e103a2e13

**Pull request merged**